core/crypto/pedersen_hash.go

```diff
@@ -1,38 +1,88 @@
+	p1 = new(starkcurve.G1Jac).FromAffine(&starkcurve.G1Affine{X: *x, Y: *y})
+
+	x.SetString("2138414695194151160943305727036575959195309218611738193261179310511854807447")
+	y.SetString("113410276730064486255102093846540133784865286929052426931474106396135072156")
+	p2 = new(starkcurve.G1Jac).FromAffine(&starkcurve.G1Affine{X: *x, Y: *y})
+
+	x.SetString("2379962749567351885752724891227938183011949129833673362440656643086021394946")
+	y.SetString("776496453633298175483985398648758586525933812536653089401905292063708816422")
+	p3 = new(starkcurve.G1Jac).FromAffine(&starkcurve.G1Affine{X: *x, Y: *y})
+}
+
+// PedersenArray implements [Pedersen array hashing].
 //
-// [Pedersen array hashing]: https://docs.starknet.io/documentation/develop/Hashing/hash-functions/#pedersen_hash
-func PedersenArray(elems ...*felt.Felt) (*felt.Felt, error) {
-	var err error
+// [Pedersen array hashing]: https://docs.starknet.io/documentation/develop/Hashing/hash-functions/#array_hashing
+func PedersenArray(elems ...*felt.Felt) *felt.Felt {
 	d := new(felt.Felt)
 	for _, e := range elems {
-		d, err = Pedersen(d, e)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	l, err := new(felt.Felt).SetInterface(len(elems))
-	if err != nil {
-		return nil, err
+		d = Pedersen(d, e)
 	}
-
-	return Pedersen(d, l)
+	return Pedersen(d, new(felt.Felt).SetUint64(uint64(len(elems))))
 }
 
-// Pedersen implements the [Pedersen hash]
+// Pedersen implements the [Pedersen hash] based on the [reference implementation].
 //
 // [Pedersen hash]: https://docs.starknet.io/documentation/develop/Hashing/hash-functions/#pedersen_hash
-func Pedersen(a *felt.Felt, b *felt.Felt) (*felt.Felt, error) {
-	out := make([]byte, starkware.BufferSize)
-	if err := starkware.Hash(a.Marshal(), b.Marshal(), out); err != nil {
-		return nil, err
-	}
-	return new(felt.Felt).SetBytes(out[:starkware.FeltSize]), nil
+// [reference implementation]: https://github.com/starkware-libs/cairo-lang/blob/de741b92657f245a50caab99cfaef093152fd8be/src/starkware/crypto/signature/fast_pedersen_hash.py
+func Pedersen(a *felt.Felt, b *felt.Felt) *felt.Felt {
+	aBytes := a.Bytes()
+	bBytes := b.Bytes()
+	aElement := new(fp.Element).SetBytes(aBytes[:])
+	bElement := new(fp.Element).SetBytes(bBytes[:])
+	result := new(starkcurve.G1Jac).Set(shiftPoint)
+	result.AddAssign(processElement(aElement, p0, p1))
+	result.AddAssign(processElement(bElement, p2, p3))
+	resultBytes := new(starkcurve.G1Affine).FromJacobian(result).X.Bytes()
+	return new(felt.Felt).SetBytes(resultBytes[:])
+}
+
+func processElement(a *fp.Element, p1 *starkcurve.G1Jac, p2 *starkcurve.G1Jac) *starkcurve.G1Jac {
+	lowPart := a.ToRegular()
+	lowPart[3] = lowPart[3] & lowPartMask // Zero-out the top nibble (bits 249-252)
+	m := new(starkcurve.G1Jac).ScalarMultiplication(p1, lowPart.ToBigInt(new(big.Int)))
+
+	highPart := new(big.Int).SetUint64(a.ToRegular()[3] >> lowPartBits & 0xf) // The top nibble (bits 249-252)
```

Comment:
`& 0xf` seem redundant here since `a` is a `felt(fp.Element)` which means that `a.ToRegular()[3] >> lowPartBits` can never be bigger than a `felt`
```suggestion
	highPart := new(big.Int).SetUint64(a.ToRegular()[3] >> lowPartBits) // The top nibble (bits 249-252)
```

Reply:
Nice catch, thanks! Fixed.